### PR TITLE
[Patch 1/2] implement AP capability report test (4.6.1)

### PR DESCRIPTION
**Pull request by CoralMalachi**
## Description

According to the Wi-Fi CERTIFIED EasyMesh™ Test Plan v1.0, test 4.6.1, which is the AP capability report test, the MAUT is triggered by the CTT controller with an AP Capability Query message, and the MAUT should reply with AP Capability Report message.

According to Multi-AP-Spesification, when a Multi-AP Agent receives an AP
Capability Query message, it shall respond within one second with an AP
Capability Report message which includes one AP Capability TLV, and one AP Radio Basic Capabilities TLV for each AP radio.

Currently, the handling of the AP Capability Query message is in the responsibility of the son slave thread, which makes no sense, since one slave has no way to know the AP capabilities of the other slaves (and should not).

The handling of the AP Capability Query message should be part of the backhaul manager, and for that reason, the following changes were made:

- Each one of the radios sends its supported channels list as part of the cACTION_BACKHAUL_ENABLE
VS CMDU to the backhaul manager.

- The backhaul manager saves for each radio its supported channels list

- When the AP Capability Query message is received in the backhaul manager, an AP Capability Report message is created, which contains one AP Capability TLV, and one AP Radio Basic Capabilities TLV for each AP radio.

## Testbed status

_Those changes were checked on the testbed with the excepted results._

### CODE REVIEW

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (tomereli) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: Important - see #691 - this will need to change to the radio mac if #431 will be merged before this change (and it probably will)

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (tomereli) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: use `int` please

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (tomereli) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: This is duplicated code - please use `add_radio_basic_capabilities` and refactor it accordingly if needed so you can use it from here.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (tomereli) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Also change the comment (in `add_radio_basic_capabilities()`) - to // TODO get maximum supported VAPs from AP Capabilities from BWL.

**Review by tomereli — DISMISSED** [review]

**Comment by morantr** [comment]
commit message headline spelling mistake 418e1dc0eba335ab740ea3af76577ecb84dcd5e4
eahc -> each

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml
Comment (arnout) [inline common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml:64]: I don't particularly like these fixed-length lists. Why not add make it a dynamic length list by adding a supported_channels_list_len before it?

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml
Comment (arnout) [inline common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml]: Now the ruid==mac change has been merged, this is no longer needed. radio_mac is available in the action header.

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/son_slave_thread.cpp:3216]: You have to check the "return value":
```
if (!std::get<0>(tuple_supported_channels)) {
    LOG(ERROR) << ...;
    do error handline
}
```

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (arnout) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: As mentioned, this will be removed.

Also, please always assign values in the same order as they are specified in the yaml file. This is essential in case there's a dynamic length list before it. It's not the case now, but it might be in the future.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h]: I think these two should be merged into a single map with a struct radio_info (that still has to be defined) as the rhs.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: See, this is annoying: we're creating a std::vector that is always the same size! Dynamic length would be better.

BTW also this is a container where the size doesn't change and it's known up-front (even in case of a dynamic length list) so use a std::array instead.

It would actually be nice if we would add a tlvf function that would return such an array for a dynamic list. It would simplify our life a lot in many places.

**Review by arnout — COMMENTED** [review]
Partial review only - I got distracted by other stuff.

I thought it would have been nicer if the slaves would create the EasyMesh TLVs and send those to the backhaul manager, which collects them in a single CMDU. Unfortunately, the way the TLV factory is built, it's quite complicated to take that approach. So this approach of storing all info on both backhaul manager and slave is OK.

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml
Comment (CoralMalachi) [reply common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml:64]: I created a fixed-length size list since I saw that in the other YAML files this supported_channels is defined as a fixed-length size list (e.g. beerocks_message_apmanager.yaml, beerocks_message_common.yaml ).
 I agree with you that it makes more sense to make it a dynamic length list, but shouldn't we do it in the rest of the place we declare this list?

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: This TODO is no longer valid.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1793]: We should add a comment here that the capability bitmask stays at 0 because we don't support unassociated STA metrics or agent-initiated steering.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: That will be difficult since both of them use private members (from slave_thread and backhaul_manager). So the best we can do is to add a function (where?) that fills in the capabilities based on a channel list.

Or maybe we should add a full-fledged class for the radio description (radio UID, channels, associated clients, and all the other info we need to cache) and use it in both backhaul manager and slave.

File: tools/docker/tests/test_flows.sh
Comment (arnout) [inline tools/docker/tests/test_flows.sh]: Please also update the controller to log the details of the AP Capability Report it receives, and check that in test_flows.

**Review by arnout — CHANGES_REQUESTED** [review]
The changes that I'm asking for are so invasive that a re-review will be required, so marking as changes requested.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (CoralMalachi) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: we also can pass those private members as arguments to this function.

Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: I've created a new class which for the moment, contains just one static method: add_ap_radio_basic_capabilities. Required private members are passed as parameters.
This class is intended to contain code to build up TLVs and to be shared between backhaul_manager_thread and son_slave_thread (or any other agent thread)

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Done

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: I've removed the ruid field

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml
Comment (mariomaz) [reply common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml:64]: Looks like `supported_channels` is treated as a fixed-length list all over the code. IMO it should not be defined as a dynamic length list.

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml
Comment (mariomaz) [reply common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml]: Done

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h]: LGTM
Is there any reason to use both operator[] and insert() to add items to the maps?
insert() does not insert if key already exists.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Removed

File: tools/docker/tests/test_flows.sh
Comment (mariomaz) [reply tools/docker/tests/test_flows.sh]: Sorry, but I need a detailed explanation on how to do this

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1793]: Done

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h]: Done

File: tools/docker/tests/test_flows.sh
Comment (ghost) [reply tools/docker/tests/test_flows.sh]: @mariomaz Did you get the answer you were looking for?

Comment (mariomaz) [reply tools/docker/tests/test_flows.sh]: No, I didn't

Comment (mariomaz) [reply tools/docker/tests/test_flows.sh]: Added handler in controller and check in test_flows.sh

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (tomereli) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: @LiorAmram Can you add such a function?

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml
Comment (tomereli) [inline common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml]: This should be a fixup commit to the first one.

File: common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_backhaul.h
Comment (tomereli) [inline common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_backhaul.h]: Same for this commit - should be squashed to the original one in this series.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (tomereli) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Same here - another fixup commit.

File: agent/src/beerocks/slave/agent_tlvf_utils.cpp
Comment (tomereli) [inline agent/src/beerocks/slave/agent_tlvf_utils.cpp]: This commit's purpose is to remove duplicate code which is great. The commit message is also great.
It's hilarious that the commit message itslef is duplicated twice though 🤣

File: agent/src/beerocks/slave/agent_tlvf_utils.cpp
Comment (tomereli) [inline agent/src/beerocks/slave/agent_tlvf_utils.cpp]: File name should be tlvf_utils.cpp (its in the agent's path so its clear it is for the agent).
Also, since you created this file - you can also add a Copyright of you and your company.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (tomereli) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: I like this!

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (tomereli) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1794]: I guess this commit should also be a fixup

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (tomereli) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Also a fixup

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h
Comment (tomereli) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h]:
```suggestion
    struct sRadioInfo {
```

Comment (tomereli) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h]: Please add doxygen comment on this new structure

File: tests/test_flows.sh
Comment (tomereli) [inline tests/test_flows.sh:206]: Should be a fixup to the commit which added the test and just reorder it to be last

**Review by tomereli — APPROVED** [review]
Most my comments are regarding squashing / fixups - but they are important, so please do your best to organize and squash commits so the whole change can be "digested" by a first time reviewer.
If this is too much work to do it all - then do what you can.
Other than that and some more small comments, LGTM - I think after some cleanups it should be good for merge, so approving.

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml
Comment (mariomaz) [reply common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml]: Fixed

File: common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_backhaul.h
Comment (mariomaz) [reply common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_backhaul.h]: Fixed

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Fixed

File: agent/src/beerocks/slave/agent_tlvf_utils.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/agent_tlvf_utils.cpp]: Fixed

File: agent/src/beerocks/slave/agent_tlvf_utils.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/agent_tlvf_utils.cpp]: Renamed. 
I think I don't deserve any credit for this code since I haven't created it, just brought it from another file as is.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1794]: Fixed

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h]: Fixed

Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h]: Fixed

File: tests/test_flows.sh
Comment (mariomaz) [reply tests/test_flows.sh:206]: Fixed

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Fixed

**Comment by mariomaz** [comment]
Commits on PR "Feature/add associated clients tlv to topology response #786" where pushed on top of the commits of this PR and they have already been merged so this PR can be closed.

**Pull request closed without merging**
